lib/livebook/file_system/git.ex

```diff
@@ -0,0 +1,195 @@
+defmodule Livebook.FileSystem.Git do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  alias Livebook.FileSystem
+
+  # File system backed by an Git repository.
+
+  @type t :: %__MODULE__{
+          id: String.t(),
+          repo_url: String.t(),
+          branch: String.t(),
+          key: String.t(),
+          external_id: String.t() | nil,
+          hub_id: String.t()
+        }
+
+  embedded_schema do
+    field :repo_url, :string
+    field :branch, :string
+    field :key, :string, redact: true
+    field :external_id, :string
+    field :hub_id, :string
+  end
+
+  @doc """
+  Returns an `%Ecto.Changeset{}` for tracking file system changes.
+  """
+  @spec change_file_system(t(), map()) :: Ecto.Changeset.t()
+  def change_file_system(git, attrs \\ %{}) do
+    changeset(git, attrs)
+  end
+
+  defp changeset(git, attrs) do
+    git
+    |> cast(attrs, [:repo_url, :branch, :key, :external_id, :hub_id])
+    |> validate_format(:repo_url, ~r/^git@[\w\.\-]+:(?:v\d+\/)?[\w\.\-\/]+\.git$/,
+      message: "must be a valid repo URL"
+    )
+    |> validate_required([:repo_url, :branch, :key, :hub_id])
+    |> update_change(:key, fn key ->
+      if key =~ "\n" do
+        key
+      else
+        String.replace(
+          key,
+          ~r/ (?!(?:OPENSSH |RSA |EC |DSA |PRIVATE )?(?:PRIVATE )?KEY-----)/,
+          "\n"
+        )
+      end
+    end)
+    |> put_id()
+  end
+
+  defp put_id(changeset) do
+    hub_id = get_field(changeset, :hub_id)
+    repo_url = get_field(changeset, :repo_url)
+
+    if get_field(changeset, :id) do
+      changeset
+    else
+      put_change(changeset, :id, FileSystem.Utils.id("git", hub_id, repo_url))
+    end
+  end
+
+  @doc false
+  def git_dir(%__MODULE__{id: id}), do: Path.join(Livebook.Config.tmp_path(), id)
+
+  @doc false
+  def ssh_path, do: Path.join(Livebook.Config.home(), ".ssh")
```

Comment:
Given that we write the key to a file temporarily and then immediately remove, is there a reason we write to livebook home rather than tmp? Is it for visibility in case we fail to remove it?

Reply:
It was for visibility during tests, but doesn't sound needed anymore